File: worker/src/counter.ts
/**
 * Counter Durable Object.
 *
 * A single global instance (`COUNTER.idFromName("global")`) keeps:
 *
 * - cumulative totals ({ requests, searches, fetches, errors, startedAt }),
 *   persisted to Durable Object storage on every update, and
 * - the last {@link RING_SIZE} request events in an in-memory ring that is
 *   never persisted (it is live telemetry, not durable history).
 *
 * The worker talks to it over internal `fetch()` requests and it exposes a
 * WebSocket endpoint that streams each new event to connected clients.
 *
 * Concurrency: Durable Objects serialize event delivery and gate storage
 * writes, so read-modify-write on the in-memory totals is safe as long as the
 * mutation is synchronous and the `put()` is awaited before yielding.
 */

/** A single observed request. Mirrored by the worker in `index.ts`. */
export type RequestEvent = {
  id: string;
  ts: number;
  method: string;
  path: string;
  kind: "search" | "fetch" | "other";
  status: number;
  ms: number;
  /** First 8 hex chars of a salted hash of the client IP. */
  client: string;
};

/** Cumulative, durable counters. */
export type Totals = {
  requests: number;
  searches: number;
  fetches: number;
  errors: number;
  startedAt: number;
};

/** How many recent events the live ring keeps. */
export const RING_SIZE = 30;

const TOTALS_KEY = "totals";

function defaultTotals(): Totals {
  return { requests: 0, searches: 0, fetches: 0, errors: 0, startedAt: Date.now() };
}

function json(data: unknown, status = 200): Response {
  return new Response(JSON.stringify(data), {
    status,
    headers: { "content-type": "application/json; charset=utf-8" },
  });
}

function normalizeEvent(value: unknown): RequestEvent {
  if (typeof value !== "object" || value === null) {
    throw new Error("event must be an object");
  }
  const record = value as Record<string, unknown>;
  const kind = record.kind;
  if (kind !== "search" && kind !== "fetch" && kind !== "other") {
    throw new Error("event.kind must be search|fetch|other");
  }
  return {
    id: typeof record.id === "string" ? record.id : crypto.randomUUID(),
    ts: typeof record.ts === "number" ? record.ts : Date.now(),
    method: typeof record.method === "string" ? record.method : "GET",
    path: typeof record.path === "string" ? record.path : "/",
    kind,
    status: typeof record.status === "number" ? record.status : 0,
    ms: typeof record.ms === "number" ? record.ms : 0,
    client: typeof record.client === "string" ? record.client : "unknown",
  };
}

export class Counter {
  private readonly state: DurableObjectState;
  private totals: Totals | null = null;
  private readonly events: RequestEvent[] = [];
  private readonly sockets = new Set<WebSocket>();

  constructor(state: DurableObjectState) {
    this.state = state;
    // Load (or initialize) totals before handling any request.
    state.blockConcurrencyWhile(async () => {
      const stored = await state.storage.get<Totals>(TOTALS_KEY);
      this.totals = stored ?? defaultTotals();
      if (!stored) {
        await state.storage.put(TOTALS_KEY, this.totals);
      }
    });
  }

  async fetch(request: Request): Promise<Response> {
    // A WebSocket upgrade passes straight through to the socket handler.
    if (request.headers.get("Upgrade")?.toLowerCase() === "websocket") {
      return this.handleWebSocket();
    }

    const { pathname } = new URL(request.url);
    switch (pathname) {
      case "/add":
        return this.handleAdd(request);
      case "/state":
        return json({ totals: await this.getTotals(), live: this.events });
      case "/totals":
        return json(await this.getTotals());
      case "/events":
        return json(this.events);
      default:
        return json({ error: "not found" }, 404);
    }
  }

  /** Read the current totals (in-memory; loaded once at construction). */
  async getTotals(): Promise<Totals> {
    if (!this.totals) {
      this.totals = (await this.state.storage.get<Totals>(TOTALS_KEY)) ?? defaultTotals();
    }
    return this.totals;
  }

  /** Record one event: bump totals, push to the ring, persist, broadcast. */
  async addEvent(event: RequestEvent): Promise<void> {
    const totals = await this.getTotals();
    totals.requests += 1;
    if (event.kind === "search") {
      totals.searches += 1;
    } else if (event.kind === "fetch") {
      totals.fetches += 1;
    }
    if (event.status >= 400) {
      totals.errors += 1;
    }

    this.events.push(event);
    if (this.events.length > RING_SIZE) {
      this.events.splice(0, this.events.length - RING_SIZE);
    }

    await this.state.storage.put(TOTALS_KEY, totals);
    this.broadcast(JSON.stringify({ type: "event", event }));
  }

  private async handleAdd(request: Request): Promise<Response> {
    let event: RequestEvent;
    try {
      event = normalizeEvent(await request.json());
    } catch (error) {
      return json({ error: "invalid event", detail: (error as Error).message }, 400);
    }
    await this.addEvent(event);
    return json({ ok: true });
  }

  private handleWebSocket(): Response {
    const pair = new WebSocketPair();
    const client = pair[0];
    const server = pair[1];
    server.accept();
    this.sockets.add(server);

    const cleanup = () => {
      this.sockets.delete(server);
    };
    server.addEventListener("close", cleanup);
    server.addEventListener("error", cleanup);
    server.addEventListener("message", (message) => {
      if (message.data === "ping") {
        try {
          server.send("pong");
        } catch {
          this.sockets.delete(server);
        }
      }
    });

    // Prime the client with the current totals and live ring.
    server.send(
      JSON.stringify({
        type: "hello",
        totals: this.totals ?? defaultTotals(),
        live: this.events,
      }),
    );

    return new Response(null, { status: 101, webSocket: client });
  }

  private broadcast(payload: string): void {
    for (const socket of this.sockets) {
      try {
        socket.send(payload);
      } catch {
        this.sockets.delete(socket);
      }
    }
  }
}

File: worker/src/counter.rs
//! Counter Durable Object: durable totals + a live in-memory event ring and
//! a WebSocket stream of new events.
//!
//! One global instance. Totals persist to storage on every add; the ring and
//! the socket set are in-memory (the DO stays alive while sockets are open).

use serde::{Deserialize, Serialize};
use serde_json::json;
use worker::*;

const RING_SIZE: usize = 30;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct RequestEvent {
    pub id: String,
    pub ts: u64,
    pub method: String,
    pub path: String,
    pub kind: String,
    pub status: u16,
    pub ms: u64,
    pub client: String,
    pub tier: String,
}

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct Totals {
    pub requests: u64,
    pub searches: u64,
    pub fetches: u64,
    pub errors: u64,
    #[serde(alias = "startedAt")]
    pub started_at: u64,
}

#[durable_object]
pub struct Counter {
    state: State,
    ring: std::cell::RefCell<Vec<RequestEvent>>,
    sockets: std::cell::RefCell<Vec<WebSocket>>,
    totals: std::cell::RefCell<Option<Totals>>,
}

impl DurableObject for Counter {
    fn new(state: State, _env: Env) -> Self {
        Self {
            state,
            ring: std::cell::RefCell::new(Vec::new()),
            sockets: std::cell::RefCell::new(Vec::new()),
            totals: std::cell::RefCell::new(None),
        }
    }

    async fn fetch(&self, mut req: Request) -> Result<Response> {
        let url = req.url()?;
        let path = url.path().to_string();
        match (req.method(), path.as_str()) {
            (Method::Post, "/add") => {
                let event: RequestEvent = serde_json::from_str(&req.text().await?)
                    .map_err(|e| Error::RustError(e.to_string()))?;
                self.add(event).await?;
                Response::from_json(&json!({ "ok": true }))
            }
            (Method::Get, "/state") => {
                let totals = self.load_totals().await?;
                let live = self.ring.borrow().clone();
                Response::from_json(&json!({ "totals": totals, "live": live }))
            }
            (Method::Get, "/stream") => {
                let pair = WebSocketPair::new()?;
                pair.server.accept()?;
                self.sockets.borrow_mut().push(pair.server.clone());
                Response::from_websocket(pair.client)
            }
            _ => Response::from_json(&json!({ "error": "not found" })).map(|r| r.with_status(404)),
        }
    }
}

impl Counter {
    async fn load_totals(&self) -> Result<Totals> {
        if let Some(totals) = self.totals.borrow().clone() {
            return Ok(totals);
        }
        let stored = self
            .state
            .storage()
            .get::<Totals>("totals")
            .await?
            .unwrap_or_default();
        *self.totals.borrow_mut() = Some(stored.clone());
        Ok(stored)
    }

    async fn add(&self, event: RequestEvent) -> Result<()> {
        let mut totals = self.load_totals().await?;
        totals.requests += 1;
        if event.kind == "search" {
            totals.searches += 1;
        }
        if event.kind == "fetch" {
            totals.fetches += 1;
        }
        if event.status >= 400 {
            totals.errors += 1;
        }
        if totals.started_at == 0 {
            totals.started_at = event.ts;
        }
        self.state.storage().put("totals", &totals).await?;
        *self.totals.borrow_mut() = Some(totals);

        {
            let mut ring = self.ring.borrow_mut();
            ring.push(event.clone());
            let overflow = ring.len().saturating_sub(RING_SIZE);
            if overflow > 0 {
                ring.drain(0..overflow);
            }
        }

        let payload = json!({ "type": "event", "event": event }).to_string();
        self.sockets
            .borrow_mut()
            .retain(|socket| socket.send_with_str(&payload).is_ok());
        Ok(())
    }
}
